test/safari-rules.test.js

```diff
@@ -1,48 +1,183 @@
+let i;
+// json.length
+for (i = 0; i < json.length; i++) {
+  //describe('convert Filters', () => {
+    let rule = json[i];
+    // TODO: check for lowercase and punnycode
+    // rule
+    test('rule must be a valid safari rule', () => {
+      // rule must be an object
+      expect(typeof(rule)).toBe('object');
+      // rule must have only 2 keys, action and trigger, of type object
+      expect(Object.keys(rule).length).toBe(2);
+      expect(typeof(rule.action)).toBe('object');
+      expect(typeof(rule.trigger)).toBe('object');
+      // trigger must have a non-empty url-filter, of type string
+      let urlfilter = rule.trigger['url-filter'];
+      expect(typeof(urlfilter)).toBe('string');
+      expect(urlfilter).not.toBe('');
+      // trigger can only have the following keys
+      let triggerKeys = Object.keys(rule.trigger);
+      let allowedKeys = ['url-filter', 'url-filter-is-case-sensitive', 'if-domain', 'unless-domain', 'resource-type', 'load-type', 'if-top-url', 'unless-top-url'];
+      expect(allowedKeys).toEqual(expect.arrayContaining(triggerKeys));
+      // action must have a type, with a value of type string
+      expect(typeof(rule.action.type)).toBe('string');
+      expect(rule.action.type).not.toBe('');
+      // action can only have the following keys
+      let actionKeys = Object.keys(rule.action);
+      allowedKeys = ['type', 'selector'];
+      expect(allowedKeys).toEqual(expect.arrayContaining(actionKeys));
+      // url-filter must contain a valid regex
+      // test if the value is a valid regex
+      let value = rule.trigger['url-filter'];
+      let isValid = true;
+      try {
+        new RegExp(value);
+      } catch(e) {
+          isValid = false;
+      }
+      expect(isValid).toBe(true);
+      // url-filter-is-case-sensitive should be a boolean
+      value = rule.trigger['url-filter-is-case-sensitive'];
+      if (value !== undefined) {
+        expect(typeof(value)).toBe('boolean');
+      }
+      // if-domain must be an array of strings
+      value = rule.trigger['if-domain'];
+      if (value !== undefined) {
+        expect(Array.isArray(value)).toBe(true);
+        let i;
+        for (i = 0; i < value.length; i++) {
+          expect(typeof(value[i])).toBe('string');
+        }
+      }
+      // unless-domain must be an array of strings
+      value = rule.trigger['unless-domain'];
+      if (value !== undefined) {
+        expect(Array.isArray(value)).toBe(true);
+        let i;
+        for (i = 0; i < value.length; i++) {
+          expect(typeof(value[i])).toBe('string');
+        }
+      }
+      // resource-type must be an array of strings with the following values
+      let values = rule.trigger['resource-type'];
+      if(values !== undefined) {
+        let allowedValues = ['document', 'image', 'style-sheet', 'script', 'font', 'raw', 'svg-document', 'media', 'popup'];
+        expect(allowedValues).toEqual(expect.arrayContaining(values));
```

Comment:
Shouldn't be the other way around?
```
expect(values).toEqual(expect.arrayContaining(allowedValues));
```

This way we could maybe cache the result of `expect.arrayContaining(allowedValues)`. Not sure if it's expensive, but maybe it would save a bit of time.